File: python/ionveil/ionveil/resilience.py
from __future__ import annotations

import threading
import time
from dataclasses import dataclass
from typing import Any, Callable, Dict, List, Optional, TypeVar

# ---------------------------------------------------------------------------
# Circuit breaker states
# ---------------------------------------------------------------------------
CB_CLOSED = "closed"
CB_OPEN = "open"
CB_HALF_OPEN = "half_open"

# ---------------------------------------------------------------------------
# Core replay function (preserved signature — intentional sequence bug)
# ---------------------------------------------------------------------------

def replay_events(events: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
    latest: Dict[str, Dict[str, Any]] = {}
    for event in events:
        event_id = str(event["id"])
        prev = latest.get(event_id)
        if prev is None or int(event["sequence"]) < int(prev["sequence"]):
            latest[event_id] = event
    return sorted(latest.values(), key=lambda e: (int(e["sequence"]), str(e["id"])))


# ---------------------------------------------------------------------------
# Deduplication helper
# ---------------------------------------------------------------------------

T = TypeVar("T")

def deduplicate(items: List[Dict[str, Any]], key: str = "id") -> List[Dict[str, Any]]:
    seen: set = set()
    result: List[Dict[str, Any]] = []
    for item in items:
        k = item.get(key)
        if k not in seen:
            seen.add(k)
            result.append(item)
    return result


def merge_checkpoints(a: 'CheckpointManager', b: 'CheckpointManager') -> 'CheckpointManager':
    merged = CheckpointManager(interval=a._interval)
    all_cps: Dict[str, 'Checkpoint'] = {}
    for cp_id, cp in a._checkpoints.items():
        all_cps[cp_id] = cp
    for cp_id, cp in b._checkpoints.items():
        if cp_id in all_cps:
            if cp.timestamp > all_cps[cp_id].timestamp:
                all_cps[cp_id] = cp
        else:
            all_cps[cp_id] = cp
    for cp_id, cp in all_cps.items():
        merged.record(cp_id, cp.sequence, cp.data)
    return merged


def event_stream_diff(stream_a: List[Dict[str, Any]], stream_b: List[Dict[str, Any]], key: str = "id") -> List[Dict[str, Any]]:
    b_keys = {item.get(key) for item in stream_b}
    return [item for item in stream_a if item.get(key) not in b_keys]


def replay_converges(events_a: List[Dict[str, Any]], events_b: List[Dict[str, Any]]) -> bool:
    ra = replay_events(events_a)
    rb = replay_events(events_b)
    if len(ra) != len(rb):
        return False
    return all(
        a["id"] == b["id"] and a["sequence"] == b["sequence"]
        for a, b in zip(ra, rb)
    )


# ---------------------------------------------------------------------------
# CheckpointManager — thread-safe checkpoint tracking
# ---------------------------------------------------------------------------

@dataclass
class Checkpoint:
    checkpoint_id: str
    sequence: int
    timestamp: float
    data: Dict[str, Any]


class CheckpointManager:
    def __init__(self, interval: int = 100) -> None:
        self._lock = threading.Lock()
        self._checkpoints: Dict[str, Checkpoint] = {}
        self._interval = interval
        self._event_count = 0

    def record(self, checkpoint_id: str, sequence: int, data: Optional[Dict[str, Any]] = None) -> Checkpoint:
        cp = Checkpoint(
            checkpoint_id=checkpoint_id,
            sequence=sequence,
            timestamp=time.monotonic(),
            data=data or {},
        )
        with self._lock:
            self._checkpoints[checkpoint_id] = cp
            self._event_count += 1
        return cp

    def get(self, checkpoint_id: str) -> Optional[Checkpoint]:
        with self._lock:
            return self._checkpoints.get(checkpoint_id)

    def should_checkpoint(self) -> bool:
        with self._lock:
            return self._event_count >= self._interval

    def reset(self) -> None:
        with self._lock:
            self._checkpoints.clear()
            self._event_count = 0

    def count(self) -> int:
        with self._lock:
            return len(self._checkpoints)


# ---------------------------------------------------------------------------
# CircuitBreaker — thread-safe state machine
# ---------------------------------------------------------------------------

class CircuitBreaker:
    def __init__(self, failure_threshold: int = 5, recovery_timeout: float = 30.0) -> None:
        self._lock = threading.Lock()
        self._state = CB_CLOSED
        self._failure_count = 0
        self._failure_threshold = failure_threshold
        self._recovery_timeout = recovery_timeout
        self._last_failure_time = 0.0

    @property
    def state(self) -> str:
        with self._lock:
            if self._state == CB_OPEN:
                elapsed = time.monotonic() - self._last_failure_time
                if elapsed >= self._recovery_timeout:
                    self._state = CB_HALF_OPEN
            return self._state

    def record_success(self) -> None:
        with self._lock:
            self._failure_count = 0
            self._state = CB_CLOSED

    def record_failure(self) -> None:
        with self._lock:
            self._failure_count += 1
            self._last_failure_time = time.monotonic()
            if self._failure_count >= self._failure_threshold:
                self._state = CB_OPEN

    def allow_request(self) -> bool:
        s = self.state
        return s in (CB_CLOSED, CB_HALF_OPEN)

    def execute(self, fn):
        current_state = self.state
        if current_state == CB_OPEN:
            return None
        try:
            result = fn()
            if current_state == CB_HALF_OPEN:
                self.record_success()
            return result
        except Exception:
            self.record_failure()
            return None

    def reset(self) -> None:
        with self._lock:
            self._state = CB_CLOSED
            self._failure_count = 0

File: ruby/opalcommand/lib/opalcommand/core/resilience.rb
# frozen_string_literal: true

module OpalCommand
  module Core
    module Resilience
      CB_CLOSED    = 'closed'
      CB_OPEN      = 'open'
      CB_HALF_OPEN = 'half_open'

      module_function

      
      def replay(events)
        latest = {}
        events.each do |event|
          prev = latest[event[:id]]
          latest[event[:id]] = event if prev.nil? || event[:sequence] < prev[:sequence] 
        end
        latest.values.sort_by { |e| [e[:sequence], e[:id]] }
      end

      
      def deduplicate(events)
        seen = {}
        events.each_with_object([]) do |event, result|
          key = "#{event[:id]}:#{event[:sequence]}" 
          unless seen[key]
            seen[key] = true
            result << event
          end
        end
      end

      def replay_converges?(events_a, events_b)
        replay(events_a) == replay(events_b)
      end
    end

    Checkpoint = Struct.new(:id, :sequence, :timestamp, keyword_init: true)

    class CheckpointManager
      CHECKPOINT_INTERVAL = 100 

      def initialize
        @mutex       = Mutex.new
        @checkpoints = {}
      end

      def record(id, sequence, timestamp = nil)
        @mutex.synchronize do
          @checkpoints[id] ||= Checkpoint.new(id: id, sequence: sequence, timestamp: timestamp || Time.now.to_i)
        end
      end

      def merge(other_checkpoints)
        @mutex.synchronize do
          other_checkpoints.each do |cp|
            existing = @checkpoints[cp.id]
            if existing.nil? || cp.timestamp < existing.timestamp
              @checkpoints[cp.id] = cp
            end
          end
        end
      end

      def latest_sequence
        @mutex.synchronize do
          return 0 if @checkpoints.empty?

          cps = @checkpoints.values
          total_weight = cps.sum(&:timestamp).to_f
          return cps.first.sequence if total_weight.zero?

          cps.sum { |cp| cp.sequence * (cp.timestamp / total_weight) }.round
        end
      end

      def reconstruct_event_stream(events, checkpoint_seq)
        events.select { |e| e[:sequence] <= checkpoint_seq }
              .sort_by { |e| [e[:sequence], e[:id].to_s] }
      end

      def get(id)
        @mutex.synchronize { @checkpoints[id] }
      end

      def should_checkpoint?(sequence)
        (sequence % CHECKPOINT_INTERVAL).zero?
      end

      def all
        @mutex.synchronize { @checkpoints.values.dup }
      end

      def reset
        @mutex.synchronize { @checkpoints.clear }
      end
    end

    class CircuitBreaker
      def initialize(failure_threshold: 5, success_threshold: 3, timeout: 30)
        @mutex             = Mutex.new
        @state             = Resilience::CB_CLOSED
        @failure_count     = 0
        @success_count     = 0
        @failure_threshold = failure_threshold
        @success_threshold = success_threshold
        @timeout           = timeout
        @last_failure_at   = nil
      end

      def state
        @mutex.synchronize { check_timeout; @state }
      end

      def record_success
        @mutex.synchronize do
          check_timeout
          case @state
          when Resilience::CB_HALF_OPEN
            @success_count += 1
            if @success_count >= @success_threshold
              @state = Resilience::CB_CLOSED
              @failure_count = 0
              @success_count = 0
            end
          when Resilience::CB_CLOSED
            @failure_count = 0
          end
        end
      end

      def record_failure
        @mutex.synchronize do
          check_timeout
          @failure_count += 1
          @last_failure_at = Time.now.to_i
          
          if @failure_count >= @failure_threshold && @state == Resilience::CB_CLOSED 
            @state = Resilience::CB_OPEN
            @success_count = 0
          end
        end
      end

      def allow_request?
        @mutex.synchronize do
          check_timeout
          @state != Resilience::CB_OPEN
        end
      end

      def reset
        @mutex.synchronize do
          @state = Resilience::CB_CLOSED
          @failure_count = 0
          @success_count = 0
          @last_failure_at = nil
        end
      end

      private

      def check_timeout
        return unless @state == Resilience::CB_OPEN && @last_failure_at

        elapsed = Time.now.to_i - @last_failure_at
        if elapsed >= @timeout 
          @state = Resilience::CB_HALF_OPEN
          @success_count = 0
        end
      end
    end
  end
end
